### ISSUE TITLE
`Promise.reduce()` broken in bluebird v2

### ISSUE THREAD

**Issue by overlookmotel** [TimBeyer/cls-bluebird#12]
`Promise.reduce()` loses CLS context with bluebird v2.

Example:

``` js
var cls = require('continuation-local-storage');
var ns = cls.createNamespace('test');

var Promise = require('bluebird');
require('cls-bluebird')(ns);

ns.run(function() {
    ns.set('id', 1);
    Promise.reduce([1, 2, 3], function(total, v) {
        console.log('In context 1:', ns.get('id'));
    }, 0);
});

ns.run(function() {
    ns.set('id', 2);
    Promise.reduce([1, 2, 3], function(total, v) {
        console.log('In context 2:', ns.get('id'));
    }, 0);
});
```

Expected output is:

```
In context 1: 1
In context 1: 1
In context 1: 1
In context 2: 2
In context 2: 2
In context 2: 2
```

But in reality you get:

``` js
In context 1: 1
In context 1: 1
In context 1: 1
In context 2: 1
In context 2: 1
In context 2: 1
```

So CLS context is getting mixed up between the two `reduce` callbacks.

**Comment by overlookmotel** [comment]
Here's the fix:

``` js
shimmer.wrap(Promise, 'reduce', function(reduce) {
    return function(arr, fn, value) {
        return reduce.call(this, arr, ns.bind(fn), value);
    };
});

shimmer.wrap(Promise.prototype, 'reduce', function(reduce) {
    return function(fn, value) {
        return reduce.call(this, ns.bind(fn), value);
    };
});
```

**Comment by overlookmotel** [comment]
Fixed in v2.0.0
